File: sdk/python/rcf_core/corpus.py
# NOTICE: This file is protected under RCF-PL
# [RCF:PROTECTED]
"""
Corpus — the reference background for surprisal (RCF-CORRELATION.md §4.2, §8.2).

§4.2 defines P_nat(f) as "the probability that an *independent* implementation
produces feature f, estimated over a large reference corpus." §8.2 names the
corpus as the one missing piece: "P_nat(f) needs a corpus. Proof quality equals
corpus quality."

This module builds that corpus. The estimator is **document frequency** — the
fraction of independent function units in which feature f appears at all:

    P_nat(f) = (# units containing f) / (total units)

That is the "IDF" in the §4.3 "TF-IDF for code graphs" analogy, and it matches
the spec wording exactly: an estimate of how often an independent unit *produces*
f. Counting presence (not multiplicity) is deliberate — we ask whether the world
re-invents the feature, not how heavily one author leans on it.

A frozen corpus carries the `alphabet_hash` it was built under. Two things are
only comparable under the same Σ (RCF-SIGMA.md §7), so `load_corpus` refuses a
corpus whose alphabet_hash disagrees with the live Σ — a drift tripwire, exactly
like `load_sigma`.

HONEST LIMIT (§8.2, and the "growth ring" doctrine): a corpus built from Python
stdlib measures what is banal *in Python*. That is the first ring — an
intra-language surprisal. Cross-language P_nat is a later ring over the same
interface. We never present a single-language corpus as the whole truth.
"""

from __future__ import annotations

import ast
import json
import os
import sysconfig
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

from .correlate import WeightFn, surprisal_weight_from_corpus
from .normalize_python import normalize_python
from .sigma import Sigma, SigmaError, load_sigma
from .wl import wl_features

_DEFAULT_CORPUS_PATH = Path(__file__).resolve().parent / "data" / "p_nat.json"


def _safe_join_within(base: Path, rel: str) -> Path:
    """
    Resolve `rel` under `base` and refuse anything that escapes `base` — a
    path-traversal guard for paths coming from a scanner result (dynamic input).
    Mirrors the constraint already used in rcf_cli.scanner._load_rcfignore.
    """
    base = base.resolve()
    p = (base / rel).resolve()
    if base != p and base not in p.parents:
        raise ValueError(f"path escapes project root: {rel!r}")
    return p


# ─── splitting a module into independent function units ───────────────────────

def iter_function_units(source: str) -> list[str]:
    """
    Split a Python module into per-function source units.

    Every `def`/`async def` (top-level OR nested) becomes its own unit, because
    a "method" in RCF terms is a function's logic. Returning each function's own
    source (via ast.get_source_segment) keeps a unit self-contained: a nested
    helper is counted once in its own right, not folded into its parent.

    Files that do not parse are skipped (returns []), never guessed.
    """
    try:
        tree = ast.parse(source)
    except (SyntaxError, ValueError):
        return []

    units: list[str] = []
    for node in ast.walk(tree):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            seg = ast.get_source_segment(source, node)
            if seg:
                units.append(seg)
    return units


# ─── the corpus object ────────────────────────────────────────────────────────

@dataclass(frozen=True)
class Corpus:
    """A frozen background distribution: feature -> in how many units it appeared."""

    alphabet_hash: str
    sigma_version: str
    total_units: int
    doc_freq: dict[str, int]

    def p_nat(self) -> dict[str, float]:
        """P_nat(f) = doc_freq(f) / total_units. Empty corpus -> empty map."""
        if self.total_units == 0:
            return {}
        n = float(self.total_units)
        return {f: c / n for f, c in self.doc_freq.items()}

    def weight_fn(self, floor: float = 1e-9) -> WeightFn:
        """
        The surprisal weight w(f) = -log P_nat(f) for this corpus. Reuses the
        canonical builder in correlate.py — the formula has exactly one home.
        A feature unseen in the corpus is maximally surprising (floor -> big w).
        """
        return surprisal_weight_from_corpus(self.p_nat(), floor=floor)

    def weight_threshold(self, percentile: float = 0.5) -> float:
        """
        A corpus-RELATIVE banality cut: the `percentile`-th weight among the
        corpus's own features. A unit feature lighter than this is "as common as
        the typical feature" → banal; heavier → rarer-than-typical → signature.

        Why relative: at scale, deep WL features are combinatorially many and each
        rare, so a FIXED P_nat cut marks almost everything "unique" (ratio
        saturates). Anchoring banality to the corpus's own weight distribution
        re-separates banal from rare regardless of corpus size.

        percentile ∈ [0,1]. 0.5 = median feature weight. Empty corpus → 0.0.
        """
        if not self.doc_freq:
            return 0.0
        w = self.weight_fn()
        weights = sorted(w(f) for f in self.doc_freq)
        idx = min(len(weights) - 1, max(0, int(percentile * (len(weights) - 1))))
        return weights[idx]


def build_corpus(
    unit_sources: Iterable[str],
    sigma: Sigma | None = None,
    *,
    iterations: int = 2,
) -> Corpus:
    """
    Build a Corpus from an iterable of *independent* function-unit sources.

    Each unit is lowered to a PDG and reduced to its SET of WL features (presence,
    not count) so document frequency counts units, not feature multiplicity.
    """
    sigma = sigma or load_sigma()
    doc_freq: dict[str, int] = {}
    total = 0

    for src in unit_sources:
        try:
            pdg = normalize_python(src, sigma)
        except (SyntaxError, ValueError, SigmaError):
            continue
        feats = wl_features(pdg, iterations)
        if not feats:
            continue
        total += 1
        for f in feats:  # iterate keys -> presence, dedup per unit
            doc_freq[f] = doc_freq.get(f, 0) + 1

    return Corpus(
        alphabet_hash=sigma.alphabet_hash,
        sigma_version=sigma.version,
        total_units=total,
        doc_freq=doc_freq,
    )


# ─── freeze / load (alphabet-bound, like load_sigma) ──────────────────────────

def freeze(corpus: Corpus, path: str | Path | None = None) -> Path:
    """Write a corpus to JSON. Creates the parent dir if needed."""
    # Resolve and constrain the target: a corpus is only ever a .json file. This
    # normalizes any dynamic input before it reaches a filesystem write.
    p = (Path(path) if path is not None else _DEFAULT_CORPUS_PATH).resolve()
    if p.suffix.lower() != ".json":
        raise SigmaError(f"corpus target must be a .json file, got: {p.name}")
    p.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "alphabet_hash": corpus.alphabet_hash,
        "sigma_version": corpus.sigma_version,
        "total_units": corpus.total_units,
        "doc_freq": corpus.doc_freq,
    }
    p.write_text(json.dumps(payload, sort_keys=True, separators=(",", ":")), encoding="utf-8")
    return p


def load_corpus(
    path: str | Path | None = None,
    *,
    sigma: Sigma | None = None,
    verify_alphabet: bool = True,
) -> Corpus:
    """
    Load a frozen corpus and (by default) refuse it if its alphabet_hash does not
    match the live Σ — a feature means something different under a different Σ.
    """
    p = (Path(path) if path is not None else _DEFAULT_CORPUS_PATH).resolve()
    if p.suffix.lower() != ".json":
        raise SigmaError(f"corpus source must be a .json file, got: {p.name}")
    if not p.is_file():
        raise SigmaError(f"corpus not found at: {p}")

    try:
        data = json.loads(p.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        raise SigmaError(f"corpus is not valid JSON: {e}") from e

    for key in ("alphabet_hash", "sigma_version", "total_units", "doc_freq"):
        if key not in data:
            raise SigmaError(f"corpus missing required key: {key!r}")

    if verify_alphabet:
        live = (sigma or load_sigma()).alphabet_hash
        if data["alphabet_hash"] != live:
            raise SigmaError(
                "corpus alphabet_hash mismatch — rebuild the corpus for this Σ.\n"
                f"  corpus: {data['alphabet_hash']}\n"
                f"  live  : {live}"
            )

    return Corpus(
        alphabet_hash=data["alphabet_hash"],
        sigma_version=data["sigma_version"],
        total_units=int(data["total_units"]),
        doc_freq=dict(data["doc_freq"]),
    )


# ─── stdlib source collection (the "banal in the world" half of the background) ─

def collect_stdlib_sources(limit: int | None = None) -> list[str]:
    """
    Collect per-function source units from the current Python's standard library.

    Walks the stdlib directory (sysconfig 'stdlib' path), reads .py files, and
    splits each into function units. Excludes test suites and site-packages so
    the background is "ordinary library code", not third-party noise.

    `limit` caps the number of UNITS for quick runs. When a cap truncates, the
    caller is expected to log it — silent truncation reads as "covered all" when
    it did not (no-silent-caps).
    """
    stdlib_dir = sysconfig.get_paths().get("stdlib")
    if not stdlib_dir or not os.path.isdir(stdlib_dir):
        return []
    base = Path(stdlib_dir).resolve()

    units: list[str] = []
    for root, dirs, files in os.walk(base):
        # prune noise: tests, site-packages, dunder dirs
        dirs[:] = [
            d for d in dirs
            if d not in ("test", "tests", "site-packages", "__pycache__")
            and not d.startswith("test_")
        ]
        for fname in files:
            if not fname.endswith(".py"):
                continue
            try:
                fpath = (Path(root) / fname).resolve()
                if base != fpath and base not in fpath.parents:
                    continue  # stay within the stdlib tree
                src = fpath.read_text(encoding="utf-8", errors="ignore")
            except (ValueError, OSError):
                continue
            for unit in iter_function_units(src):
                units.append(unit)
                if limit is not None and len(units) >= limit:
                    return units
    return units


def collect_project_background_sources(root: str) -> list[str]:
    """
    Collect function units from a project's NON-protected files — "ordinary code
    by the same author". Protected files are EXCLUDED on purpose: folding the
    method into its own background would teach the background the method's rarity
    and cancel its surprisal (RCF-SIGMA self-poisoning).

    Reuses rcf_cli.scanner for the is_protected verdict.
    """
    from rcf_cli.scanner import RCFScanner

    base = Path(root).resolve()
    scanner = RCFScanner(root)
    units: list[str] = []
    for res in scanner.scan_directory(include_protected=True):
        if res.get("is_protected"):
            continue  # background is the *un*protected world only
        if not res["path"].endswith(".py"):
            continue
        try:
            fpath = _safe_join_within(base, res["path"])  # reject path traversal
            src = fpath.read_text(encoding="utf-8", errors="ignore")
        except (ValueError, OSError):
            continue
        units.extend(iter_function_units(src))
    return units


# ─── CLI: build & freeze the corpus (the artifact is generated, not committed) ─

def main(argv: list[str] | None = None) -> int:
    import argparse
    import time

    parser = argparse.ArgumentParser(
        prog="python -m rcf_core.corpus",
        description="Build & freeze the P_nat surprisal corpus (stdlib + a project's non-protected code).",
    )
    parser.add_argument("--project", metavar="ROOT", help="also fold this project's NON-protected .py as background")
    parser.add_argument("--limit", type=int, default=None, help="cap stdlib units (omit = full stdlib)")
    parser.add_argument("--out", metavar="JSON", help="output path (default: rcf_core/data/p_nat.json)")
    args = parser.parse_args(argv)

    sigma = load_sigma()
    t0 = time.time()
    print("collecting stdlib sources...")
    stdlib = collect_stdlib_sources(limit=args.limit)
    if args.limit is not None and len(stdlib) >= args.limit:
        print(f"  NOTE: hit --limit={args.limit} — corpus is TRUNCATED, not exhaustive")
    print(f"  stdlib units: {len(stdlib)}  ({time.time() - t0:.1f}s)")

    proj: list[str] = []
    if args.project:
        proj = collect_project_background_sources(args.project)
        print(f"  project background units (non-protected): {len(proj)}")

    corpus = build_corpus(stdlib + proj, sigma)
    print(f"  corpus: {corpus.total_units} units, {len(corpus.doc_freq)} features, Σ {corpus.sigma_version}")

    path = freeze(corpus, args.out)
    print(f"frozen to: {path}  ({os.path.getsize(path)} bytes)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sdk/typescript/src/core/corpus.ts
// NOTICE: This file is protected under RCF-PL
// [RCF:PROTECTED]

import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs';
import path from 'path';
import Parser from 'tree-sitter';
// @ts-ignore
import TypeScript from 'tree-sitter-typescript';
import { PDG } from './pdg.js';
import { Sigma, loadSigma, SigmaError, timingSafeHashEqual } from './sigma.js';

import { wlFeatures } from './wl.js';
import { normalizeTypescript } from './normalize_typescript.js';
import { normalizeByExtension } from './normalize.js';
import { Scanner } from './Scanner.js';

export function safeJoinWithin(basePath: string, rel: string): string {
  const base = path.resolve(basePath);
  const p = path.resolve(base, rel);
  if (base !== p && !p.startsWith(base + path.sep)) {
    throw new Error(`path escapes project root: ${rel}`);
  }
  return p;
}

export function iterFunctionUnits(source: string): string[] {
  const parser = new Parser();
  parser.setLanguage(TypeScript.typescript);
  let tree;
  try {
    tree = parser.parse(source);
  } catch (e) {
    return [];
  }

  const units: string[] = [];
  function visit(node: Parser.SyntaxNode) {
    if (
      node.type === 'function_declaration' ||
      node.type === 'arrow_function' ||
      node.type === 'function_expression' ||
      node.type === 'method_definition'
    ) {
      units.push(node.text);
    }
    for (let i = 0; i < node.childCount; i++) {
      visit(node.child(i)!);
    }
  }
  if (tree.rootNode) {
    visit(tree.rootNode);
  }
  return units;
}

export class Corpus {
  public readonly alphabetHash: string;
  public readonly sigmaVersion: string;
  public readonly totalUnits: number;
  public readonly docFreq: Record<string, number>;

  constructor(
    alphabetHash: string,
    sigmaVersion: string,
    totalUnits: number,
    docFreq: Record<string, number>
  ) {
    this.alphabetHash = alphabetHash;
    this.sigmaVersion = sigmaVersion;
    this.totalUnits = totalUnits;
    this.docFreq = docFreq;
  }

  public pNat(): Record<string, number> {
    if (this.totalUnits === 0) {
      return {};
    }
    const n = this.totalUnits;
    const p: Record<string, number> = {};
    for (const [f, c] of Object.entries(this.docFreq)) {
      p[f] = c / n;
    }
    return p;
  }

  public weightFn(floor: number = 1e-9): (feature: string) => number {
    const pNatMap = this.pNat();
    return (feature: string): number => {
      const p = pNatMap[feature] !== undefined ? pNatMap[feature] : 0.0;
      return -Math.log(Math.max(p, floor));
    };
  }

  public weightThreshold(percentile: number = 0.5): number {
    const keys = Object.keys(this.docFreq);
    if (keys.length === 0) {
      return 0.0;
    }
    const w = this.weightFn();
    const weights = keys.map(f => w(f));
    weights.sort((a, b) => a - b);
    const idx = Math.min(weights.length - 1, Math.max(0, Math.floor(percentile * (weights.length - 1))));
    return weights[idx];
  }
}

export function buildCorpus(
  unitSources: Iterable<string>,
  sigma?: Sigma,
  iterations: number = 2
): Corpus {
  const s = sigma || loadSigma();
  const docFreq: Record<string, number> = {};
  let total = 0;

  for (const src of unitSources) {
    let pdg: PDG;
    try {
      pdg = normalizeTypescript(src, s);
    } catch {
      continue;
    }
    const feats = wlFeatures(pdg, iterations);
    if (Object.keys(feats).length === 0) {
      continue;
    }
    total++;
    for (const f of Object.keys(feats)) {
      docFreq[f] = (docFreq[f] || 0) + 1;
    }
  }

  return new Corpus(s.alphabetHash, s.version, total, docFreq);
}

export function freezeCorpus(corpus: Corpus, filePath: string): string {
  const resolvedPath = path.resolve(filePath);
  if (path.extname(resolvedPath).toLowerCase() !== '.json') {
    throw new SigmaError(`corpus target must be a .json file, got: ${path.basename(resolvedPath)}`);
  }
  const parent = path.dirname(resolvedPath);
  mkdirSync(parent, { recursive: true });

  const payload = {
    alphabet_hash: corpus.alphabetHash,
    sigma_version: corpus.sigmaVersion,
    total_units: corpus.totalUnits,
    doc_freq: corpus.docFreq,
  };
  writeFileSync(resolvedPath, JSON.stringify(payload, null, 2), 'utf-8');
  return resolvedPath;
}

export function loadCorpus(
  filePath: string,
  sigma?: Sigma,
  verifyAlphabet: boolean = true
): Corpus {
  const resolvedPath = path.resolve(filePath);
  if (path.extname(resolvedPath).toLowerCase() !== '.json') {
    throw new SigmaError(`corpus source must be a .json file, got: ${path.basename(resolvedPath)}`);
  }
  if (!existsSync(resolvedPath)) {
    throw new SigmaError(`corpus not found at: ${resolvedPath}`);
  }

  let data: any;
  try {
    data = JSON.parse(readFileSync(resolvedPath, 'utf-8'));
  } catch (e: any) {
    throw new SigmaError(`corpus is not valid JSON: ${e.message}`);
  }

  for (const key of ['alphabet_hash', 'sigma_version', 'total_units', 'doc_freq']) {
    if (data[key] === undefined) {
      throw new SigmaError(`corpus missing required key: '${key}'`);
    }
  }

  if (verifyAlphabet) {
    const live = (sigma || loadSigma()).alphabetHash;
    if (!timingSafeHashEqual(data.alphabet_hash ?? '', live)) {
      throw new SigmaError(
        `corpus alphabet_hash mismatch — rebuild the corpus for this Σ.`
      );
    }
  }

  return new Corpus(
    data.alphabet_hash,
    data.sigma_version,
    data.total_units,
    data.doc_freq
  );
}

export function collectProjectBackgroundSources(root: string): string[] {
  const scanner = new Scanner();
  const results = scanner.scanDirectory(root);
  const units: string[] = [];

  for (const res of results) {
    if (res.isProtected) {
      continue;
    }
    const ext = path.extname(res.path).toLowerCase();
    if (!['.ts', '.js', '.tsx', '.jsx', '.go', '.rs'].includes(ext)) {
      continue;
    }
    let src: string;
    try {
      const fpath = safeJoinWithin(root, path.relative(root, res.path));
      src = readFileSync(fpath, 'utf-8');
    } catch {
      continue;
    }
    units.push(...iterFunctionUnits(src));
  }
  return units;
}

/**
 * Build a Corpus from a list of file paths.
 * Each file is read from disk and dispatched to the correct normalizer
 * based on its extension (.go → Go, .rs → Rust, .ts/.js → TypeScript).
 */
export function buildCorpusFromFiles(
  filePaths: string[],
  sigma?: Sigma,
  iterations: number = 2
): Corpus {
  const s = sigma || loadSigma();
  const docFreq: Record<string, number> = {};
  let total = 0;

  for (const fpath of filePaths) {
    let src: string;
    try {
      src = readFileSync(fpath, 'utf-8');
    } catch {
      continue;
    }
    const ext = path.extname(fpath).toLowerCase();
    let pdg: PDG;
    try {
      pdg = normalizeByExtension(src, ext, s);
    } catch {
      continue;
    }
    const feats = wlFeatures(pdg, iterations);
    if (Object.keys(feats).length === 0) continue;
    total++;
    for (const f of Object.keys(feats)) {
      docFreq[f] = (docFreq[f] || 0) + 1;
    }
  }

  return new Corpus(s.alphabetHash, s.version, total, docFreq);
}
